synapse/logging/opentracing.py

```diff
@@ -641,57 +641,16 @@ def extract_text_map(carrier):
 # Tracing decorators
 
 
-def trace(func):
+def trace(func=None, operation_name=None):
     """
     Decorator to trace a function.
     Sets the operation name to that of the function's.
     """
-    if opentracing is None:
-        return func
-
-    @wraps(func)
-    def _trace_inner(self, *args, **kwargs):
-        if opentracing is None:
-            return func(self, *args, **kwargs)
-
-        scope = start_active_span(func.__name__)
-        scope.__enter__()
-
-        try:
-            result = func(self, *args, **kwargs)
-            if isinstance(result, defer.Deferred):
-
-                def call_back(result):
-                    scope.__exit__(None, None, None)
-                    return result
-
-                def err_back(result):
-                    scope.span.set_tag(tags.ERROR, True)
-                    scope.__exit__(None, None, None)
-                    return result
 
-                result.addCallbacks(call_back, err_back)
+    if func and not operation_name:
```

Comment:
does this work correctly for the `@trace / def func` case? I think this should probably move inside `decorator`.

Reply:
It works but there's less redundancy if it's inside.

Reply:
It works fine but it's a little clearer and less redundant inside decorator. I've moved it inside